crates/ra_mbe/src/subtree_source.rs

```diff
@@ -41,16 +43,18 @@ impl<'a> SubtreeTokenSource<'a> {
     }
 
     fn mk_token(&self, pos: usize) -> Token {
-        match self.get(pos) {
-            Some(tt) => Token { kind: tt.kind, is_jointed_to_next: tt.is_joint_to_next },
+        let idx = self.get(pos);
+        let cached = self.cached.borrow();
+        match cached[idx] {
+            Some(ref tt) => Token { kind: tt.kind, is_jointed_to_next: tt.is_joint_to_next },
             None => Token { kind: EOF, is_jointed_to_next: false },
         }
     }
 
-    fn get(&self, pos: usize) -> Option<TtToken> {
+    fn get(&self, pos: usize) -> usize {
```

Comment:
Would returning an `Option<Ref<TtToken>>` work here?

Reply:
I'm not sure; can I return a `Ref<T>` out of a `RefCell<Vec<T>>`?

Reply:
Yes, there's `Ref::map` method: https://doc.rust-lang.org/std/cell/struct.Ref.html#method.map

Reply:
TIL. But I don't think it works, as `get` has a `RefMut`, and there seems to be no way to downgrade a `RefMut` to a `Ref`. I could probably drop it and borrow again, but it's not zero-cost.

Reply:
I *think* it should be effectively zero cost, as LLVM should be able to collapse two operations on the same "number of borrows" int. So, I still prefer using `Ref` to indixes here, as it seems clearer.

Reply:
Made it a `Ref<Option<TtToken>>` instead, but it should be good enough :sweat_smile:.

Unfortunately I can't confirm that the borrows go away, as the assembly is now a mess (perhaps because of the incremental build).